airflow/migrations/versions/0003_3_0_0_remove_is_subdag.py

```
@@ -0,0 +1,76 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""
+Remove SubDAGs: ``is_subdag`` & ``root_dag_id`` columns from DAG table.
+
+Revision ID: d0f1c55954fa
+Revises: 044f740568ec
+Create Date: 2024-08-11 21:32:40.576172
+
+"""
+
+from __future__ import annotations
+
+import sqlalchemy as sa
+from alembic import op
+
+from airflow.migrations.db_types import StringID
+
+# revision identifiers, used by Alembic.
+revision = "d0f1c55954fa"
+down_revision = "044f740568ec"
+branch_labels = None
+depends_on = None
+airflow_version = "3.0.0"
+
+
+def _column_exists(inspector, column_name):
+    return column_name in [col["name"] for col in inspector.get_columns("dag")]
+
+
+def _index_exists(inspector, index_name):
+    return index_name in [index["name"] for index in inspector.get_indexes("dag")]
+
+
+def upgrade():
+    """Remove ``is_subdag`` column from DAGs table."""
+    conn = op.get_bind()
+    inspector = sa.inspect(conn)
+
+    with op.batch_alter_table("dag", schema=None) as batch_op:
+        if _index_exists(inspector, "idx_root_dag_id"):
+            batch_op.drop_index("idx_root_dag_id")
+        if _column_exists(inspector, "is_subdag"):
```

Why are we doing these conditionally? How could the columns not exist?

---

I didn't realize most of our migration aren't idempotent. I had a first pass at those few years back : https://github.com/apache/airflow/pulls?q=is%3Apr+idempotent+migration+is%3Aclosed+author%3Akaxil

This is great for cases where downgrades or upgrades don't complete fully. You can rerun run.

---

I see. That's really only an issue on our old friend mysql, who lacks transactional DDL 🙃.

---

You could make the argument that, in particular when adding, this "if not exists" idempotency is actually opening up the possibility for inconsistency. What if the column type changed later? Probably better to let it fail if the db is in an inconsistent state.

---

Column type shouldn't matter though since we are just checking if the column exists or not in that condition -- rest all remains same.